### DISCUSSION TITLE
Trying to use stbi_load_from_memory to obtain texture file from a glb file.

### DISCUSSION THREAD

**Discussion by dc2032** [nothings/stb#1369] [Q&A]
The NASA webite has fantastic 3D models of all nine planets, which I have downloaded. I am trying to develop a game for my 14 year old grandson where you fly a spaceship through the solar system. The models are downloaded in GLTF format, with a PNG file imbedded. I have successfully used cgltf.h to load all of the model data into memory, including the PNG file. I am trying to build an openGL texture file using stbi_load_from_memory. I'm calling it like this: 
unsigned char* data = stbi_load_from_memory((unsigned char*)image, size, &width, &height, &channels, 0);
On return from the call, in which I do not get any error messages, data points to a buffer containing all xCC. Single stepping through the Visual Studio 2022 debugger, the buffer appears to get corrupted in stbi__parse_png_file, but I have been unable to trap the exact point. I know the PNG file is valid, because I can access the buffer and write it out to a file and then read it using Windows Photo viewer. All I'm looking for is some guidance on what could be wrong with the file in memory and perhaps what I could do about it.

**Comment by nothings** [comment]
if you write it out to a file, can you then load it with stbi_load? if that corrupts in exactly the same way, please post that file.

**Reply by dc2032** [reply]
Yes, it corrupts as well. The resulting file is attached.
![jupiter_diff](https://user-images.githubusercontent.com/113209175/189484330-f15c262b-5e42-4dfc-929a-aae23ca84db8.png)

**Reply by PathogenDavid** [reply]
@dc2032 stb_image loads that file just fine on my end. It might be that GitHub re-encoded it when you uploaded it, can you zip it and upload that instead?

**Reply by dc2032** [reply]
Here is the file zipped. Thanks for your help.
[jupiter_diff.zip](https://github.com/nothings/stb/files/9540566/jupiter_diff.zip)

**Reply by PathogenDavid** [reply]
Unfortunately turns out GitHub didn't re-encode it, that file's hash is identical and loads fine:

![Screenshot of the Jupiter texture on a cube](https://user-images.githubusercontent.com/278957/189493339-e2e5d797-bbfc-4528-9770-641ec8634dc5.png)

Any chance you're also using assimp somewhere else in the project? You might be running into this issue: https://github.com/nothings/stb/discussions/1364

**Reply by dc2032** [reply]
I had installed assimp as suggested by Joey de Vries in his book Learn OpenGL - Graphics Programming, but I never used it in this project. In fact I never used it at all because his sample code also used three other libraries I was not using. I did a search though, to make sure there were no includes I may have put in, but no such luck. I am using Visual Studio 2022 on Windows 11, so I'm in the process of porting my gltfloader program to my linux box to see if it works there. I'll let you know how I make out. Thanks.

**Reply by nothings** [reply]
I double-checked loading that image with stbi_load and with stbi_load_from_memory, both are fine. Are you using the most current version of stb_image?

I've stepped through stbi__parse_png, and it doesn't seem to be doing anything remarkable, just an awful lot of 8KB IDAT chunks (since the file is 10MB).

Are you compiling for 64-bit or for 32-bit?

**Reply by dc2032** [reply]
The version of stb_image I have is 2.27 and I've been compiling for 64-bit.

**Reply by dc2032** [reply]
I decided to try the solution suggested in the discussion David referred to earlier and put in the #define STB_IMAGE_STATIC. But when I do that I get unresolved references to stb_load_from_memory and stb_load_from_file. I'm also not making much headway trying it on linux. It appears I have a version problem, gcc is looking for glibc 2.27 but I have 2.28 installed. I'm still researching that one.

**Reply by nothings** [reply]
For reference, I doubt STB_IMAGE_STATIC is the issue, but it's a tool that allows multiple instances of stb_image in the same project without name collisions. If you use it, it makes the functions private to the c/cpp file with the implementation, so you have to write your own wrappers in that file to allow it to be called externally.

**Reply by nothings** [reply]
Also, you should try loading the file-written image yourself in the same program just using stbi_load and see if it has the same problem or not. If that works ok, then try stbi_load_from_memory but loading it from the file. If that's ok, then you should triangulate what's different in the call that works and the call that fails--are you using the wrong 'size', etc.

All of this should be straightforward and simpler and more direct to figuring out what's going wrong in stb_image than trying to build another program on another operating system.

**Reply by dc2032** [reply]
When I try the file, I get the same corruption. At one point I had converted the PNG file produced to a JPG and I just tried stbi_load with it and had the same problem. It looks like I may have a Visual Studio problem because sometimes as I'm single stepping in the debugger, it will exit my program without generating an exception.
